boa/src/syntax/lexer/mod.rs

```diff
@@ -229,6 +223,225 @@ impl<'a> Lexer<'a> {
         }
     }
 
+    /// Lexes a numerical literal.
+    ///
+    /// More information:
+    ///  - [ECMAScript Specification][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-literals-numeric-literals
+    fn reed_numerical_literal(&mut self, ch: char) -> Result<(), LexerError> {
+        /// This is a helper structure
+        ///
+        /// This structure helps with identifying what numerical type it is and what base is it.
+        enum NumericKind {
+            Rational,
+            Integer(u32),
+            BigInt(u32),
+        }
+
+        impl NumericKind {
+            /// Get the base of the number kind.
+            fn base(&self) -> u32 {
+                match self {
+                    Self::Rational => 10,
+                    Self::Integer(ref base) => *base,
+                    Self::BigInt(ref base) => *base,
+                }
+            }
+
+            /// Converts `self` to BigInt kind.
+            fn convert_to_bigint(&mut self) {
+                *self = match *self {
+                    Self::Rational => unreachable!("can not convert rational number to BigInt"),
+                    Self::Integer(base) => Self::BigInt(base),
+                    Self::BigInt(base) => Self::BigInt(base),
+                };
+            }
+        }
+
+        // TODO: Setup strict mode.
+        let strict_mode = false;
```

Comment:
Currently I have `strict_mode` set to always be `false`. In strict mode decimals with leading zeros and implicit octal are not allowed.

I seems like the parser should take in a lexer, because `'use strict';` it only applies to function or program scope and the lexer does not know what a scope is.

Reply:
Yep, this is going to be interesting to see once we have the lexer as an iterator tied to the cursor.